[SYCL] Remove discard_events queue property

JackAKirk opened the pull request:
Apparently this is already deprecated in favour of the sycl_ext_oneapi_enqueue_functions extension.
The main consequence of this is that users of discard_events will have to ensure they use the new `submit_without_event` / `nd_launch` etc APIs from https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_enqueue_functions.asciidoc

This should leave all other functionality working correctly.

igchor commented:
@JackAKirk can you also cleanup event_impl? I still see some references to discard events there. Or is this still needed for some paths?

JackAKirk commented:
> @JackAKirk can you also cleanup event_impl? I still see some references to discard events there. Or is this still needed for some paths?

I've removed one case that is definitely safe to remove. `waitInternal` could also potentially be cleaned up, although this depends on the way forward with host_task; in general I think that all of the internal discard event checks could be removed following on from https://github.com/intel/llvm/pull/18018 . Apart from host_task the only other reason for keeping `LastEvent` is to keep this little potential optimization: https://github.com/intel/llvm/pull/18018/files#diff-f412476f167a41ca486ea41ab76bc7e861df86e8cf0542d4931b93947f41e6edL687

However as discussed this doesn't seem worth it.
You do still have a `LastHostTaskEvent` in the https://github.com/intel/llvm/pull/18018 draft, so maybe it makes sense to wait until this PR is finalized/merged to finish the internal discard_event cleanup.
Could you get rid of `LastHostTaskEvent` completely, and instead submit a empty barrier to the native queue after the host_task is completed? This would guarantee that the host_task is synced with the in-order queue, and if users want to use multiple in order queues they can `submit_with_event` the host_task, and sync manually?
This would mean that `sycl::queue::host_task` matches exactly the semantics of `cudaLaunchHostFunc(cudaStream_t, ...)`, see
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#host-functions-callbacks

igchor commented:
> > @JackAKirk can you also cleanup event_impl? I still see some references to discard events there. Or is this still needed for some paths?
> 
> I've removed one case that is definitely safe to remove. `waitInternal` could also potentially be cleaned up, although this depends on the way forward with host_task; in general I think that all of the internal discard event checks could be removed following on from #18018 . Apart from host_task the only other reason for keeping `LastEvent` is to keep this little potential optimization: https://github.com/intel/llvm/pull/18018/files#diff-f412476f167a41ca486ea41ab76bc7e861df86e8cf0542d4931b93947f41e6edL687
> 
> However as discussed this doesn't seem worth it. You do still have a `LastHostTaskEvent` in the #18018 draft, so maybe it makes sense to wait until this PR is finalized/merged to finish the internal discard_event cleanup. Could you get rid of `LastHostTaskEvent` completely, and instead submit a empty barrier to the native queue after the host_task is completed? This would guarantee that the host_task is synced with the in-order queue, and if users want to use multiple in order queues they can `submit_with_event` the host_task, and sync manually? This would mean that `sycl::queue::host_task` matches exactly the semantics of `cudaLaunchHostFunc(cudaStream_t, ...)`, see https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#host-functions-callbacks

Yes, the way you suggest would be the best option - I believe @KseniyaTikhomirova is working on a similar approach. However, there are some issues related to memory residence with this approach so this will probably happen as a next step.

JackAKirk commented on sycl/include/sycl/properties/queue_properties.def:
```diff
@@ -9,8 +9,6 @@
 __SYCL_DATA_LESS_PROP(property::queue, in_order, InOrder)
 __SYCL_DATA_LESS_PROP(property::queue, enable_profiling, QueueEnableProfiling)
 
-__SYCL_DATA_LESS_PROP(ext::oneapi::property::queue, discard_events,
```
I've now kept the property but made it a no-op. I've updated the PR description accordingly.
I've kept all discard events tests other than those which test that the underlying `sycl::event` is `nullptr`, this this PR makes those tests obsolete.

gmlueck commented on sycl/doc/extensions/deprecated/sycl_ext_oneapi_discard_queue_events.asciidoc:
```diff
@@ -0,0 +1,211 @@
+== Status
+
+This extension has been deprecated.  Although the interfaces defined in this
+specification are still supported in {dpcpp}, we expect that they will be
```
You need to define `dpcpp` as an asciidoc attribute at the top of the file.

gmlueck replied:
I saw you fixed the `dpcpp` in the other asciidoc file, but this one is not yet fixed.

JackAKirk replied:
Done

JackAKirk replied:
> I saw you fixed the `dpcpp` in the other asciidoc file, but this one is not yet fixed.

Does this line not fix it? https://github.com/intel/llvm/pull/18059/files#diff-2bb29c57c2d8ec658778edf583e9084912e6f13dbebe26508c7b2da012738158R11

It seems to render "DPC++" correctly.

gmlueck commented on sycl/doc/extensions/deprecated/sycl_ext_oneapi_discard_queue_events.asciidoc:
```diff
@@ -0,0 +1,211 @@
+extension: see link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[here].
+*Shipping software products should stop using APIs defined in this
+specification and use this alternative instead.*
+
```
I think you could add some words here noting that this extension no longer provides any benefit, and direct people to the enqueue function extension as an alternative.

JackAKirk replied:
I've added the sentence "This extension no longer provides any benefit" before the existing explanation and sentence pointing to the enqueue_functions extension. I've also copied this to the doc in the supported folder. Let me know if you think I should change it any further.
Thanks

gmlueck reviewed (commented):
The title of the PR says that it deprecates the `discard_events` property, but I don't see anything in this PR that adds a deprecation warning.  Was that done in a previous PR?  The deprecation warning should direct people to use the [sycl_ext_oneapi_enqueue_functions](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_enqueue_functions.asciidoc) extension instead.

JackAKirk commented:
> The title of the PR says that it deprecates the `discard_events` property, but I don't see anything in this PR that adds a deprecation warning. Was that done in a previous PR? The deprecation warning should direct people to use the [sycl_ext_oneapi_enqueue_functions](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_enqueue_functions.asciidoc) extension instead.

Thanks, yeah I completely missed this. I've added the warning now.

kbenzie reviewed (approved):
sycl/test-e2e/Adatpers LGTM

JackAKirk commented:
- I've moved the remaining DiscardEvents test to the DeprecatedFeatures folder.
- Added a test for the deprecation warning to sycl/test/warnings/warnings_deprecated.cpp
- Removed duplicate testing for discard_events queue property from other tests: the remaining tests in DiscardEvents folder should be sufficient to check that basic functionality remains with the discard_events property.

jbrodman commented:
Looks like there's a conflict to resolve?

ProGTX commented:
> Looks like there's a conflict to resolve?

Conflicts resolved. The CI seems a bit flaky, Windows testing passed after a few reruns.

Also friendly ping @intel/unified-runtime-reviewers-level-zero

vinser52 commented:
@JackAKirk what's next with this PR?

JackAKirk commented on sycl/source/handler.cpp:
```diff
@@ -505,8 +505,8 @@ event handler::finalize() {
-      bool DiscardEvent = (MQueue->MDiscardEvents || !impl->MEventNeeded) &&
-                          MQueue->supportsDiscardingPiEvents();
+      bool DiscardEvent =
+          !impl->MEventNeeded && MQueue->supportsDiscardingPiEvents();
```
Thanks, yeah I see this now: sycl_ext_oneapi_enqueue_functions.asciidoc doesn't require an in-order queue.
Fixed

JackAKirk replied:
Although note that this change is technically out of the scope of just removing discard_events queue property functionality: The change on this line now makes more discarding of events in the impl for `submit_without_event` on a sycl::queue without the in-order property. This seems to make some tests fail, despite it being allowed by the spec doc as I mention above. I will investigate.

JackAKirk commented:
> @JackAKirk what's next with this PR?

This should be ready to merge once tests pass.

JackAKirk commented:
Test failures appear unrelated and occur in other PRs: see https://github.com/intel/llvm/issues/18416

JackAKirk commented:
@intel/llvm-gatekeepers this is ready to be merged.

Thanks
